core/src/main/java/org/apache/accumulo/core/fate/Fate.java

```diff
@@ -73,9 +75,12 @@ public class Fate<T> {
   private final FateStore<T> store;
   private final T environment;
   private final ScheduledThreadPoolExecutor fatePoolWatcher;
-  private final ExecutorService executor;
+  private final ExecutorService transactionExecutor;
+  private final ExecutorService deadResCleanerExecutor;
 
   private static final EnumSet<TStatus> FINISHED_STATES = EnumSet.of(FAILED, SUCCESSFUL, UNKNOWN);
+  private static boolean userDeadReservationCleanerRunning = false;
+  private static boolean metaDeadReservationCleanerRunning = false;
```

Comment:
> Should these be Atomic? Or is it fine to keep them as-is?

Right now, there is only one Manager which runs, so these do not need to be Atomic, but the goal is to have multiple managers in the future, so not sure if this should be changed now or in the future.

> Can these static variables could be removed? Do these help testing in someway? In the manager we can assume that only a single instance of each type is created. If we wanted to guard against multiple instances per type, then a more general check could be done that is not limited to dead reservation cleaners.

The purpose of these was to avoid creating more than 1 DeadReservationCleaner running per store type. As I currently have this, the first Fate object to be created per store type will be running a DeadReservationCleaner (in the future, this probably should be changed as running on the first Fate object created may not be the best approach). In the Manager, only one Fate object will be created per store type, so this isn't a necessary check now, but isn't the plan in the future to have multiple Fate instance running per store type, so in that case we'd only ever want 1 DeadReservationCleaner running? I was premtively handling this case. What do you have in mind regarding a more general check? Or should I not be handling this case yet?

Reply:
Left as-is until further clarification

Reply:
In #3964  I was thinking of having this [PartitionData](https://github.com/apache/accumulo/blob/208e39927310f8ff7dd5978f2794e26bee6be1b4/core/src/main/java/org/apache/accumulo/core/manager/PartitionData.java) object that would be passed to Fate from the manager.  Fate could use that to partition fate operations to run and to determine if it should run a singleton service using the `PartitionData.shouldRun` method.   Was thinking that PatitionData could be passed to all the managers high level services (like Fate, CompactionCoordinator, TabletGroupWatcher) and that all could use it to partition their data in their own way.

With something like that, in test if you create 3 fate instances then could pass them `ParitionData(0,3)`, `ParitionData(1,3)`, and `ParitionData(2,3)`.   It probably does not makes sense to add that in this PR as it will not partition the fate ids at this point.

For this PR we could simplify things and have a boolean argument on the Fate constructor that indicates if the dead reservation detection should be started and not have any static variable that attempt to enforce one per process.

Reply:
:+1: re the PartitionData object
Got it, I've changed it to take a boolean argument for whether the cleaner should run.